### PR TITLE
Fix clear prefix check to avoid erasing child trie roots.

### PR DESCRIPTION
**Pull request by cheme**
Checks on child storage root for clear prefix is incorrect, this changes it.

An alternative implementation would be to allow such clear prefix and skip the child storage root prefix from being deleted.

This is possibly breaking consensus but I know of no pallet that can  delete a prefix containing child storage root, so from my point of view it should be fine to use.

### CODE REVIEW

**Review by bkchr — COMMENTED** [review]
Could you add some test?

**Comment by cheme** [comment]
Will do :+1:

File: primitives/state-machine/src/ext.rs [hunk 1]
```diff
@@ -1022,6 +1023,24 @@ mod tests {
 			ext.child_storage_hash(child_info, &[30]),
 			Some(Blake2Hasher::hash(&[31]).as_ref().to_vec()),
 		);
+
+		// check clear prefix is ignored on conflict
+		use sp_core::storage::well_known_keys;
```
Comment (bkchr) [inline primitives/state-machine/src/ext.rs:1028]: I would prefer to make this an independent test.

File: primitives/storage/src/lib.rs [hunk 2]
```diff
@@ -180,6 +180,16 @@ pub mod well_known_keys {
 		// Other code might depend on this, so be careful changing this.
 		key.starts_with(CHILD_STORAGE_KEY_PREFIX)
 	}
+
+	/// Whether a prefix potentially contains child storage keys.
+	pub fn can_contain_child_storage_key(prefix: &[u8]) -> bool {
```
Comment (bkchr) [inline primitives/storage/src/lib.rs:185]:
```suggestion
	/// Returns if the given `key` starts with [`CHILD_STORAGE_KEY_PREFIX`] or collides with it.
	pub fn starts_with_child_storage_key(key: &[u8]) -> bool {
```

**Review by bkchr — APPROVED** [review]
Some last nitpicks, otherwise looks good.

**Review by thiolliere — APPROVED** [review]
good catch, looks good to me,

I wonder should a similar handling be done with next_storage_key ? like if next_storage_key goes into some CHILD_STORAGE_KEY_PREFIX space then it should be skipped ?

**Comment by cheme** [comment]
I would not bother to skip it, accessing root read only seems fine to me.

**Comment by thiolliere** [comment]
Yes I agree

**Comment by cheme** [comment]
bot merge

**Pull request merged**